tests/run-make/inaccessible-temp-dir/rmake.rs

```diff
@@ -0,0 +1,44 @@
+// Issue #66530: We would ICE if someone compiled with `-o /dev/null`,
+// because we would try to generate auxiliary files in `/dev/` (which
+// at least the OS X file system rejects).
+//
+// An attempt to `-Ztemps-dir` into a directory we cannot write into should
+// indeed be an error; but not an ICE.
+//
+// However, some folks run tests as root, which can write `/dev/` and end
+// up clobbering `/dev/null`. Instead we'll use an inaccessible path, which
+// also used to ICE, but even root can't magically write there.
+//
+// Note that `-Ztemps-dir` uses `create_dir_all` so it is not sufficient to
+// use a directory with non-existing parent like `/does-not-exist/output`.
+// See https://github.com/rust-lang/rust/issues/66530
+
+//@ only-linux
+// Reason: set_mode is only available on Unix
+
+//@ ignore-arm
+// Reason: linker error on `armhf-gnu`
+
+use run_make_support::{fs_wrapper, rustc};
+
+fn main() {
+    // Create an inaccessible directory.
+    fs_wrapper::create_dir("inaccessible");
+    let meta = fs_wrapper::metadata("inaccessible");
+    let mut perms = meta.permissions();
+    perms.set_mode(0o000); // Lock down the directory.
+    fs_wrapper::set_permissions("inaccessible", perms);
+
+    // Run rustc with `-Z temps-dir` set to a directory *inside* the inaccessible one,
+    // so that it can't create `tmp`.
+    rustc()
+        .input("program.rs")
+        .arg("-Ztemps-dir=inaccessible/tmp")
+        .run_fail()
+        .assert_stderr_contains(
+            "failed to find or create the directory specified by `--temps-dir`",
+        );
+
+    perms.set_mode(0o666); // Unlock the directory, so that compiletest can delete it.
```

Comment:
Curious: is `compiletest` really unable to remove this directory without these two lines? I thought that https://github.com/rust-lang/rust/blob/cfdb6175c2286acdd0c7d71a2ac9eefb777eec09/src/tools/compiletest/src/runtest.rs#L3420 takes care of that. If this is not needed, it would be nice to remove these two lines.

Reply:
I actually can't tell: from a cursory glance it might fail to remove a readonly on linux because it uses fs::remove_file which is documented to possibly fail if we don't have perms

Reply:
Let's find out. Removing the lines!

Reply:
You might need to double check the build directory for this specific test to see if the file lingers around

Reply:
I tried it and it's problematic, to say the last. Let's keep the lines :laughing:

Reply:
"To say the least"? That is all the convincing I needed. The lines have been added back.

Reply:
We should also document this

Reply:
IIRC there was another run-make test that sets the perms to readonly but does not set it back, which I think causes something to linger around as well

Reply:
There is already a comment after the first line, I think that's enough.

Reply:
> We should also document this
There is a comment that says `Unlock the directory, so that compiletest can delete it.`, which further information do you think would be good to have?

Reply:
Ah yes, none other than https://github.com/rust-lang/rust/blob/0c960618b56f662d933e8b864cd9632a99174e87/tests/run-make/lto-readonly-lib/Makefile#L10

Reply:
> There is already a comment after the first line, I think that's enough.

(Sorry, that was mostly a comment for myself, I wanted to document this as either part of the support library and/or the dev-guide)

Reply:
That's one of the other PRs I currently have open: #126095. The 444 gets enacted by this:

```rs
perms.set_readonly(true);
```

Do you think this test port should also get final lines to restore the permissions back to read/write?

Reply:
You can try locally if you can run the test twice in a row. If not, then yes, the permissions should be restored :)
Btw, seems like it could be useful to have some helper like `set_readonly(path)` in `run-make-support`.

Reply:
I gave it a shot and also removed `ignore-windows`, since it isn't using chmod anymore.